keps/sig-node/1948-add-deallocate-to-device-plugin-api/README.md

```diff
@@ -0,0 +1,217 @@
+    * If whatever is programmed on the FPGA has network access, it runs the risk of continuing to send and respond to packets and pollute the network.
+  - For dynamically binding/unbinding drivers for the devices as needed.
+- `Deallocate`:
+  - For use with complex device plugins that require tracking the state of their devices and learning when they are no longer in use. For example, multi modal devices. Multi modal devices can operate in more than one mode of operation, and thus have to be advertised by the device plugin as two separate devices, and the device plugin has to take care to stop advertising a device when its being used in the other mode, and so on. An example of multi modal devices is also using FPGAs in the following 2 modes:
+    * Use the entire FPGA as a device
+    * Split the FPGA between multiple users, essentially advertising one FPGA as multiple smaller FPGAs.
+    * A device plugin needs to know when a full FPGA will stop being used so it can go back to advertise the FPGA partitions, and vice versa.
+  - To maintain the same logical splitting of `Allocate` and `PreStartContainer`
+
+### Goals
+
+- Add the `PostStopContainer` and `Deallocate` API calls to the device plugin API.
+- Make the new added API calls optional, as they are not needed for all devices.
+- Maintain compatibility with existing device plugins.
+
+### Non-Goals
+
+- Make any modifications to the main API calls of the device plugin API.
+- Make changes specific to one type of devices.
+
+## Proposal
+
+The device plugin API includes API calls for:
+- `Allocate`: Which is used to instruct device plugins to allocate device(s) to requesting containers.
+- `PreStartContainer`: (Optional). Which allow the device plugins to do device initialization, loading drivers, and any other initialization actions that may be needed.
+
+This KEP proposes adding two extra API calls, maintaining the same logical reasoning of the previous two. Those are:
+- `Deallocate`: (Optional). Which is the opposite of allocate, and is needed to inform device plugins that some devices are no longer being used (this used to happen silently before)
+- `PostStopContainer`: (Optional). Which allow the device plugins to do device cleanup, driver unloading, and any other cleanup actions that may be needed.
+
+Since both additions are optional, existing device plugins should continue functioning properly with no needed modifications. Only device plugins that wish to utilize the new API calls will need to be modified.
+
+### Risks and Mitigations
+
+Only risk is breaking existing device plugins by introducing non optional changes. Can be mitigated by enough test coverage.
+
+## Design Details
+
+- Remove reusable devices from `devicemanage`. Reusable devices are silently reused by kubernetes, and so they can't be reported back to device plugins for cleanup, tracking, etc.
```

Comment:
From my perspective, this is the biggest blocker to adding something like this (at least with how it is currently described).

At least in the case of GPUs, we rely on this functionality to run init containers that are guaranteed to be allocated the same set of GPUs that the app containers and "warm them up" for when the app containers finally start.

Reply:
Moreover "something" like `reusable` devices will always be required to satisfy the "effective requests/limits" of a Pod as defined by:

        https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#resources

The rule states that a Pod’s "effective request/limit" for a resource should be the larger of:
        * The highest of any particular resource request or limit defined on all init Containers
        * The sum of all app Containers request/limit for a resource

I don't know how to satisfy this if we drop the notion of `reusable` devices.

There is likely a way to have both `reusable` devices *and*  the `Deallocate()` call, but I don't think we can just drop `reusable`.

Reply:
Thanks for taking the time to review.

A question though, is this "reuse" feature being used by anything other than init containers? Does kubernetes have the liberty to hold resources for longer?

If I remember correctly, we could call the PostStopContainer hook without modifying or removing the resuability. Deallocation was another story.

Reply:
> A question though, is this "reuse" feature being used by anything other than init containers? Does kubernetes have the liberty to hold resources for longer?

No. It is exclusively designed to reuse the devices allocated to `init` containers inside `app` containers of the same within the same pod. There is no "reusing" of devices across `app` containers in the same pod or across pods.

Reply:
I see. I guess I agree then, there must be a way to deallocate devices when they are no longer being reused. I will change the KEP sometime later today to reflect this.

Reply:
it doesn't imply you "have to" reuse the same devices, but it does imply that the total number of `init devices` + `app devices` must never be larger than `app devices` and the natural way to do that is to reuse the same devices.

Reply:
As an example, it should be completely valid on a machine with 4 devices to have:
```
initContainer:
    limits:
        mydevice: 4
appContainer:
    limits:
        mydevice: 4
```
and without reuse, I see no way of doing this